### PR TITLE
Extract duplicated code for collection notifications and sectioned results

### PR DESCRIPTION
**Pull request by tgoyne**
There should be no functional changes here. This is just eliminating some duplicated code that lead to me having to update more places than expected when adding notifications on actors.

No changelog entry since there shouldn't be any user-visible changes from this.

### CODE REVIEW

File: Realm/RLMResults.mm [hunk 1]
```diff
@@ -27,8 +27,8 @@
 #import "RLMObservation.hpp"
 #import "RLMProperty_Private.h"
 #import "RLMQueryUtil.hpp"
-#import "RLMRealm_Private.hpp"
 #import "RLMRealmConfiguration_Private.hpp"
+#import "RLMScheduler.h"
```
Comment (dianaafanador3) [inline Realm/RLMResults.mm:31]: Is this import even needed?

**Review by dianaafanador3 — COMMENTED** [review]
Pretty cool refactor

File: Realm/RLMCollection.mm [hunk 2]
```diff
@@ -441,7 +441,7 @@ - (NSString *)description {
 
 namespace {
 struct CollectionCallbackWrapper {
-    void (^block)(id, RLMCollectionChange *, NSError *);
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMCollection.mm:444]: There's different block types used for different collections. Obj-c doesn't really give us a way to express that other than making the property entirely dynamically typed.

File: Realm/RLMCollection.mm [hunk 3]
```diff
@@ -480,51 +480,51 @@ - (void)invalidate {
...
         }
-        NSError *error;
-        RLMRealm *realm = token->_realm = [RLMRealm realmWithConfiguration:config queue:queue error:&error];
-        if (!realm) {
-            block(nil, nil, error);
-            return;
-        }
-        RLMCollection *collection = [realm resolveThreadSafeReference:tsr];
-        token->_token = RLMGetBackingCollection(collection).add_notification_callback(CollectionCallbackWrapper{block, collection, skipFirst}, std::move(keyPathArray));
+        RLMRealm *realm = token->_realm = [RLMRealm realmWithConfiguration:config queue:queue error:nil];
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMCollection.mm:508]: Yes, there's now no way for obtaining a new Realm instance to fail if the Realm is already open.

**Review by dianaafanador3 — APPROVED** [review]
LGTM

**Pull request merged**